minimint-api/src/db/rocksdb_impl.rs

```diff
@@ -36,40 +37,88 @@ impl Database for rocksdb::OptimisticTransactionDB {
+    //     let tx = self.transaction();
+    //
+    //     for change in batch.iter() {
+    //         match change {
+    //             BatchItem::InsertNewElement(element) => {
+    //                 if tx.get(element.key.to_bytes()).unwrap().is_some() {
+    //                     tx.put(element.key.to_bytes(), element.value.to_bytes())?;
+    //                     error!("Database replaced element! This should not happen!");
+    //                     trace!("Problematic key: {:?}", element.key);
+    //                 } else {
+    //                     tx.put(element.key.to_bytes(), element.value.to_bytes())?;
+    //                 }
+    //             }
+    //             BatchItem::InsertElement(element) => {
+    //                 tx.put(element.key.to_bytes(), element.value.to_bytes())?;
+    //             }
+    //             BatchItem::DeleteElement(key) => {
+    //                 if tx.get(key.to_bytes()).unwrap().is_none() {
+    //                     tx.delete(key.to_bytes())?;
+    //                     error!("Database deleted absent element! This should not happen!");
+    //                     trace!("Problematic key: {:?}", key);
+    //                 } else {
+    //                     tx.delete(key.to_bytes())?;
+    //                 }
+    //             }
+    //             BatchItem::MaybeDeleteElement(key) => {
+    //                 tx.delete(key.to_bytes())?;
+    //             }
+    //         }
+    //     }
+    //     tx.commit()?;
+    //     Ok(())
+    // }
+
+    fn transaction<'a>(&'a self) -> Result<Box<dyn Transaction<'a>>> {
+        Ok(Box::new(self.transaction()))
+    }
+}
+
+impl Transaction for rocksdb::Transaction<DB> {
+    fn raw_insert_entry(&mut self, key: &[u8], value: Vec<u8>) -> Result<Option<Vec<u8>>> {
+        let val = self.get(key).unwrap();
+        self.put(key, value)?;
+        Ok(val)
+    }
+
+    fn raw_get_value(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        Ok(self.get(key)?)
+    }
+
+    fn raw_maybe_remove_entry(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        let val = self.get(key).unwrap();
+        self.delete(key)?;
+        Ok(val)
+    }
+
+    fn raw_find_by_prefix(&self, key_prefix: &[u8]) -> PrefixIter<'_> {
+        let prefix = key_prefix.to_vec();
+        Box::new(
+            self.prefix_iterator(prefix.clone())
+                .map_while(move |res| res.0.starts_with(&prefix).then(|| res))
+                .map(|(key_bytes, value_bytes)| (key_bytes.to_vec(), value_bytes.to_vec()))
+                .map(Ok),
+        )
+    }
+
+    fn checkpoint(self) -> Result<()> {
+        Ok(self.set_savepoint())
+    }
+
+    fn rollback(self) -> Result<(), anyhow::Error> {
+        self.rollback_to_savepoint().map_err(|e| anyhow::Error)
+    }
+
+    fn abort(self) -> Result<()> {
+        self.abort()
+    }
+
+    fn commit(self) -> Result<(), anyhow::Error> {
+        self.commit().map_err(|e| anyhow::Error)
```

Comment:
Overall structural question for my own understanding: are there multiple threads accessing the database during consensus? Do we anticipate potentially running into tx conflicts when committing to the db?

Reply:
Presently there aren't multiple threads accessing this may change in the future though I'm not completely sure. Ultimately there shouldn't be transaction conflicts but if something isn't working correctly then we should be able to handle that.

Reply:
How does it handle conflicts between subtransactions, i.e. 2 updates to the same key?

Reply:
With the proposed API the way it is, there are no subtransactions (although the current code supports this with the BatchTxs)

If you try to do 2 updates to the same key in the same transaction, since we're using Optimistic Transactions, the commit will fail. If we decide to use Transactions with locks, the second update will fail instead of the commit. 

Optimistic Transactions are likely better if we don't anticipate lots of conflicts happening since they're more efficient and don't take locks

Reply:
I see...we were thinking about using the DB to handle conflicting txs (e.g. 2 peg-outs using the same UTXO).  We need to be sure the DB will resolve the conflicts correctly and doesn't enable DoS attacks.